File: Backend/src/main/java/com/example/Backend/service/ReclamationService.java
package com.example.Backend.service;

import com.example.Backend.dto.ReclamationDTO;
import com.example.Backend.entity.Reclamation;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.Map;
import java.util.UUID;

public interface ReclamationService {

    /**
     * Create a new reclamation
     * 
     * @param userId         User ID
     * @param reclamationDTO Reclamation data
     * @return Created reclamation
     */
    Reclamation createReclamation(UUID userId, ReclamationDTO reclamationDTO);

    /**
     * Get reclamation by ID
     * 
     * @param reclamationId Reclamation ID
     * @return Reclamation entity
     */
    Reclamation getReclamationById(UUID reclamationId);

    /**
     * Get all reclamations
     * 
     * @param pageable Pagination parameters
     * @return Page of reclamations
     */
    Page<Reclamation> getAllReclamations(Pageable pageable);

    /**
     * Get reclamations by user
     * 
     * @param userId   User ID
     * @param pageable Pagination parameters
     * @return Page of user's reclamations
     */
    Page<Reclamation> getReclamationsByUser(UUID userId, Pageable pageable);

    /**
     * Get reclamations by status
     * 
     * @param status   Status
     * @param pageable Pagination parameters
     * @return Page of reclamations
     */
    Page<Reclamation> getReclamationsByStatus(String status, Pageable pageable);

    /**
     * Get reclamations by category
     * 
     * @param category Category
     * @param pageable Pagination parameters
     * @return Page of reclamations
     */
    Page<Reclamation> getReclamationsByCategory(String category, Pageable pageable);

    /**
     * Get pending reclamations
     * 
     * @param pageable Pagination parameters
     * @return Page of pending reclamations
     */
    Page<Reclamation> getPendingReclamations(Pageable pageable);

    /**
     * Get reclamations assigned to support agent
     * 
     * @param supportAgentId Support agent ID
     * @param pageable       Pagination parameters
     * @return Page of reclamations
     */
    Page<Reclamation> getReclamationsByAssignedAgent(UUID supportAgentId, Pageable pageable);

    /**
     * Assign reclamation to support agent
     * 
     * @param reclamationId  Reclamation ID
     * @param supportAgentId Support agent ID
     * @return Updated reclamation
     */
    Reclamation assignToAgent(UUID reclamationId, UUID supportAgentId);

    /**
     * Update reclamation status
     * 
     * @param reclamationId Reclamation ID
     * @param status        New status
     * @return Updated reclamation
     */
    Reclamation updateStatus(UUID reclamationId, String status);

    /**
     * Add response to reclamation
     * 
     * @param reclamationId Reclamation ID
     * @param response      Response text
     * @return Updated reclamation
     */
    Reclamation addResponse(UUID reclamationId, String response);

    /**
     * Close reclamation
     * 
     * @param reclamationId Reclamation ID
     * @param resolution    Resolution text
     * @return Updated reclamation
     */
    Reclamation closeReclamation(UUID reclamationId, String resolution);

    /**
     * Get average resolution time
     * 
     * @return Average time in hours
     */
    Double getAverageResolutionTime();

    /**
     * Get reclamation statistics
     * 
     * @return Map of counts by status
     */
    Map<String, Long> getReclamationStatistics();

    /**
     * Count pending reclamations
     * 
     * @return Count
     */
    Long countPendingReclamations();
}

File: frontend-web/src/app/core/services/reclamation.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface Reclamation {
    id: string;
    ticketNumber: string;
    user: any;
    order?: any;
    subject: string;
    description: string;
    category: string;
    priority: 'LOW' | 'MEDIUM' | 'HIGH' | 'URGENT';
    status: 'OPEN' | 'IN_PROGRESS' | 'RESOLVED' | 'CLOSED';
    assignedAgent?: any;
    response?: string;
    resolution?: string;
    attachmentUrl?: string;
    createdAt: Date;
    updatedAt: Date;
    resolvedAt?: Date;
}

export interface Page<T> {
    content: T[];
    totalElements: number;
    totalPages: number;
    size: number;
    number: number;
}

export interface ReclamationStats {
    totalTickets: number;
    openTickets: number;
    inProgressTickets: number;
    inProgress: number;
    resolved: number;
    closed: number;
    resolvedToday: number;
    averageResolutionTime: number;
}

@Injectable({
    providedIn: 'root'
})
export class ReclamationService {
    private apiUrl = `${environment.apiUrl}/reclamations`;

    constructor(private http: HttpClient) { }

    // ==================== RECLAMATION CRUD ====================

    createReclamation(reclamation: Partial<Reclamation>): Observable<Reclamation> {
        return this.http.post<Reclamation>(this.apiUrl, reclamation);
    }

    getReclamationById(id: string): Observable<Reclamation> {
        return this.http.get<Reclamation>(`${this.apiUrl}/${id}`);
    }

    // Alias for getReclamationById
    getById(id: string): Observable<Reclamation> {
        return this.getReclamationById(id);
    }

    // ==================== ADMIN QUERIES ====================

    getAllReclamations(page: number = 0, size: number = 20, sort: string = 'createdAt'): Observable<Page<Reclamation>> {
        const params = new HttpParams()
            .set('page', page.toString())
            .set('size', size.toString())
            .set('sort', sort);
        return this.http.get<Page<Reclamation>>(this.apiUrl, { params });
    }

    getMyReclamations(page: number = 0, size: number = 20): Observable<Page<Reclamation>> {
        const params = new HttpParams()
            .set('page', page.toString())
            .set('size', size.toString());
        return this.http.get<Page<Reclamation>>(`${this.apiUrl}/my-reclamations`, { params });
    }

    getReclamationsByStatus(status: string, page: number = 0, size: number = 20): Observable<Page<Reclamation>> {
        const params = new HttpParams()
            .set('page', page.toString())
            .set('size', size.toString());
        return this.http.get<Page<Reclamation>>(`${this.apiUrl}/status/${status}`, { params });
    }

    getReclamationsByCategory(category: string, page: number = 0, size: number = 20): Observable<Page<Reclamation>> {
        const params = new HttpParams()
            .set('page', page.toString())
            .set('size', size.toString());
        return this.http.get<Page<Reclamation>>(`${this.apiUrl}/category/${category}`, { params });
    }

    getPendingReclamations(page: number = 0, size: number = 20): Observable<Page<Reclamation>> {
        const params = new HttpParams()
            .set('page', page.toString())
            .set('size', size.toString());
        return this.http.get<Page<Reclamation>>(`${this.apiUrl}/pending`, { params });
    }

    getMyAssignedReclamations(paramsOrPage: number | { page?: number; size?: number } = 0, size: number = 20): Observable<Page<Reclamation>> {
        let page: number;
        let pageSize: number;

        if (typeof paramsOrPage === 'object') {
            page = paramsOrPage.page || 0;
            pageSize = paramsOrPage.size || 20;
        } else {
            page = paramsOrPage;
            pageSize = size;
        }

        const params = new HttpParams()
            .set('page', page.toString())
            .set('size', pageSize.toString());
        return this.http.get<Page<Reclamation>>(`${this.apiUrl}/assigned-to-me`, { params });
    }

    getReclamationsByAgent(agentId: string, page: number = 0, size: number = 20): Observable<Page<Reclamation>> {
        const params = new HttpParams()
            .set('page', page.toString())
            .set('size', size.toString());
        return this.http.get<Page<Reclamation>>(`${this.apiUrl}/assigned/${agentId}`, { params });
    }

    // ==================== ASSIGNMENT ====================

    assignToAgent(id: string, agentId: string): Observable<Reclamation> {
        return this.http.patch<Reclamation>(`${this.apiUrl}/${id}/assign/${agentId}`, {});
    }

    assignToMe(id: string): Observable<Reclamation> {
        return this.http.patch<Reclamation>(`${this.apiUrl}/${id}/assign-to-me`, {});
    }

    // ==================== STATUS & RESPONSE ====================

    updateStatus(id: string, status: string): Observable<Reclamation> {
        return this.http.patch<Reclamation>(`${this.apiUrl}/${id}/status`, { status });
    }

    addResponse(id: string, response: string): Observable<Reclamation> {
        return this.http.post<Reclamation>(`${this.apiUrl}/${id}/response`, { response });
    }

    closeReclamation(id: string, resolution: string): Observable<Reclamation> {
        return this.http.patch<Reclamation>(`${this.apiUrl}/${id}/close`, { resolution });
    }

    // ==================== STATISTICS ====================

    getStatistics(): Observable<ReclamationStats> {
        return this.http.get<ReclamationStats>(`${this.apiUrl}/statistics`);
    }

    getAverageResolutionTime(): Observable<number> {
        return this.http.get<number>(`${this.apiUrl}/average-resolution-time`);
    }

    getPendingCount(): Observable<number> {
        return this.http.get<number>(`${this.apiUrl}/pending/count`);
    }

    // ==================== ALIASES FOR COMPATIBILITY ====================

    // Alias for getStatistics - used by support dashboard
    getStats(): Observable<ReclamationStats> {
        return this.getStatistics();
    }

    // Flexible query method - accepts an object with page, size, status, search
    getReclamations(params: { page?: number; size?: number; status?: string; search?: string }): Observable<Page<Reclamation>> {
        let httpParams = new HttpParams()
            .set('page', (params.page || 0).toString())
            .set('size', (params.size || 20).toString());

        if (params.status) {
            return this.getReclamationsByStatus(params.status, params.page || 0, params.size || 20);
        }

        return this.http.get<Page<Reclamation>>(this.apiUrl, { params: httpParams });
    }

    // ==================== AGENT PERFORMANCE ====================

    /**
     * Get current agent's performance statistics
     */
    getMyPerformance(): Observable<AgentPerformanceStats> {
        return this.http.get<AgentPerformanceStats>(`${this.apiUrl}/my-performance`);
    }

    /**
     * Get current agent's weekly statistics
     */
    getMyWeeklyStats(): Observable<WeeklyStats[]> {
        return this.http.get<WeeklyStats[]>(`${this.apiUrl}/my-weekly-stats`);
    }

    /**
     * Get current agent's recent activities
     */
    getMyActivities(limit: number = 10): Observable<RecentActivity[]> {
        const params = new HttpParams().set('limit', limit.toString());
        return this.http.get<RecentActivity[]>(`${this.apiUrl}/my-activities`, { params });
    }
}

// Additional interfaces for agent performance
export interface AgentPerformanceStats {
    totalAssigned: number;
    totalResolved: number;
    resolvedThisMonth: number;
    resolvedToday: number;
    inProgress: number;
    openTickets: number;
    avgResolutionTimeHours: number;
}

export interface WeeklyStats {
    day: string;
    tickets: number;
    resolved: number;
}

export interface RecentActivity {
    id: string;
    action: string;
    ticketSubject: string;
    time: string;
    type: 'resolved' | 'replied' | 'assigned' | 'escalated';
}
